process/transaction/transactionCostEstimator.go

```diff
@@ -1,133 +1,139 @@
-	tce.mutExecution.Lock()
-	tce.compilePerByteCost, tce.storePerByteCost = getOperationCost(gasSchedule)
-	tce.mutExecution.Unlock()
-}
-
-func getOperationCost(gasSchedule map[string]map[string]uint64) (uint64, uint64) {
-	baseOpMap, ok := gasSchedule[core.BaseOperationCost]
-	if !ok {
-		return 0, 0
-	}
-
-	storeCost, ok := baseOpMap["StorePerByte"]
-	if !ok {
-		return 0, 0
-	}
-
-	compilerCost, ok := baseOpMap["CompilePerByte"]
-	if !ok {
-		return 0, 0
-	}
-
-	return storeCost, compilerCost
-}
-
 // ComputeTransactionGasLimit will calculate how many gas units a transaction will consume
 func (tce *transactionCostEstimator) ComputeTransactionGasLimit(tx *transaction.Transaction) (*transaction.CostResponse, error) {
 	tce.mutExecution.RLock()
 	defer tce.mutExecution.RUnlock()
 
 	txType, _ := tce.txTypeHandler.ComputeTransactionType(tx)
-	tx.GasPrice = 1
 
 	switch txType {
-	case process.MoveBalance:
+	case process.SCDeployment, process.SCInvoking, process.BuiltInFunctionCall, process.MoveBalance:
+		return tce.simulateTransactionCost(tx, txType)
+	case process.RelayedTx, process.RelayedTxV2:
+		// TODO implement in the next PR
 		return &transaction.CostResponse{
-			GasUnits: tce.feeHandler.ComputeGasLimit(tx),
-		}, nil
-	case process.SCDeployment:
-		return tce.computeScDeployGasLimit(tx)
-	case process.SCInvoking:
-		return tce.computeScCallGasLimit(tx)
-	case process.BuiltInFunctionCall:
-		return tce.computeScCallGasLimit(tx)
-	case process.RelayedTx:
-		return &transaction.CostResponse{
-			GasUnits:   0,
-			RetMessage: "cannot compute cost of the relayed transaction",
+			GasUnits:      0,
+			ReturnMessage: "cannot compute cost of the relayed transaction",
 		}, nil
 	default:
 		return &transaction.CostResponse{
-			GasUnits:   0,
-			RetMessage: process.ErrWrongTransaction.Error(),
+			GasUnits:      0,
+			ReturnMessage: process.ErrWrongTransaction.Error(),
 		}, nil
 	}
 }
-func (tce *transactionCostEstimator) computeScDeployGasLimit(tx *transaction.Transaction) (*transaction.CostResponse, error) {
-	scDeployCost := uint64(len(tx.Data)) * (tce.storePerByteCost + tce.compilePerByteCost)
-	baseCost := tce.feeHandler.ComputeGasLimit(tx)
 
-	return &transaction.CostResponse{
-		GasUnits: baseCost + scDeployCost,
-	}, nil
-}
+func (tce *transactionCostEstimator) simulateTransactionCost(tx *transaction.Transaction, txType process.TransactionType) (*transaction.CostResponse, error) {
+	tce.addMissingFieldsIfNeeded(tx)
 
-func (tce *transactionCostEstimator) computeScCallGasLimit(tx *transaction.Transaction) (*transaction.CostResponse, error) {
-	scCallGasLimit, err := tce.query.ComputeScCallGasLimit(tx)
+	res, err := tce.txSimulator.ProcessTx(tx)
 	if err != nil {
 		return &transaction.CostResponse{
-			GasUnits:   0,
-			RetMessage: err.Error(),
+			GasUnits:      0,
+			ReturnMessage: err.Error(),
+		}, nil
+	}
+
+	isMoveBalanceOk := txType == process.MoveBalance && res.FailReason == ""
```

Comment:
you should treat overpaying with gas here. not in sc processor. Also you could make some dummy accounts and you ensure that there is enough balance to execute - before simulation - if you want that as well. It is not a good thing to skip those checks.

Reply:
done.